p2panda-net/src/sync/initiate.rs

```diff
@@ -0,0 +1,173 @@
+                    engine_actor_tx
+                        .send(ToEngineActor::SyncHandshakeSuccess {
+                            peer,
+                            topic: topic.clone(),
+                        })
+                        .await
+                        .map_err(|err| {
+                            SyncError::Critical(format!(
+                                "engine_actor_tx failed sending sync handshake success: {err}"
+                            ))
+                        })?;
+
+                    continue;
+                }
+
+                // 2. Data Sync Phase.
+                // ~~~~~~~~~~~~~~~~~~~
+                let FromSync::Data(header, payload) = message else {
+                    return Err(SyncError::Critical("expected to receive only data messages from sync session in data sync phase".into()));
+                };
+
+                engine_actor_tx
+                    .send(ToEngineActor::SyncMessage {
+                        header,
+                        payload,
+                        delivered_from: peer,
+                        topic: topic.clone(),
+                    })
+                    .await
+                    .map_err(|err| {
+                        SyncError::Critical(format!(
+                            "engine_actor_tx failed sending sync message: {err}"
+                        ))
+                    })?;
+            }
+
+            Ok(())
+        })
+    };
+
+    // Run the "initiating peer" side of the sync protocol.
+    let result = sync_protocol
+        .initiate(
+            topic.clone(),
+            Box::new(&mut send),
+            Box::new(&mut recv),
+            Box::new(&mut sink),
+        )
+        .await;
+
+    // Drop the tx, so the rx in the glue task receives the closing event.
+    drop(sink);
+
+    // We're expecting the task to exit with a result soon, we're awaiting it here ..
+    let glue_task_result = glue_task_handle
+        .await
+        .map_err(|err| SyncError::Critical(format!("glue task handle failed: {err}")))?;
+
+    // .. to inform some brave developer who will read the error logs ..
+    if let Err(SyncError::Critical(err)) = &glue_task_result {
+        error!("critical error in sync protocol: {err}");
+    }
+
+    // .. and forward it further.
+    glue_task_result?;
+
+    // The same we're doing with errors coming from the sync protocol implementation itself.
+    if let Err(err) = result {
+        match &err {
+            SyncError::Critical(err) => {
+                error!("critical error in sync protocol: {err}");
+            }
+            _ => {
+                warn!("error in sync protocol: {err}");
+            }
+        }
+        return Err(err);
+    }
+
+    engine_actor_tx
+        .send(ToEngineActor::SyncDone { peer, topic })
+        .await
+        .map_err(|err| {
+            SyncError::Critical(format!("engine_actor_tx failed sending sync done: {err}"))
+        })?;
```

Comment:
We should only send a `SyncDone` when _no_ error occurred in the initiator.

Reply:
`SyncFailed` is handled in the sync manager, that makes the code a little bit hard to read though?

Reply:
It does make it a little confusing (to have `SyncDone` here but not `SyncFailed`), though I'm not sure if there's a desirable alternative; it needs to be handled by the manager because we don't interpret it as a failure in all cases (such as a connection error when we still have more reattempts to go through).

Perhaps the second best option is to include a comment in `initiate.rs` (maybe L162) which explains that `SyncFailed` is sent by the sync manager according to the current state for this specific peer-topic attempt.

Reply:
Thank you! That makes sense. I agree, let's keep it as is and I'll add a comment.